PTL setUp() does not revert pbs.conf changes

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* If anyone makes changes to pbs.conf in their tests, those changes won't be reverted before the next test is run.  This may cause subsequent tests to fail.

#### Affected Platform(s)
* All Linux

#### Cause / Analysis / Design
* Well, we just didn't revert pbs.conf in setUp

#### Solution Description
* Added a function called revert_pbsconf_startbits() in PBSTestSuite to revert pbs.conf, which now gets called from setUp()
* If run on a single host type 1 installation, the function sets all start bits to 1
* If run on a multi-node setup with moms or comms, the function sets all mom/comm start bits on those hosts to 1, and on the server host it sets all bits to 1 except the mom bit.

#### Testing logs/output
* 5 new tests, results on a single node machine (expect to see the remote mom test skipped): 
[testpbstestsuite.log](https://github.com/PBSPro/pbspro/files/2652640/testpbstestsuite.log)

* results for remote mom test:
[testremotemom.log](https://github.com/PBSPro/pbspro/files/2652643/testremotemom.log)

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bhroam commented:
If you revert anything, you will need to restart the PBS services.  You don't want to have to do that if nothing changes.  You will want to verify something has changed before you restart them.  Also, the mom gets restarted as part of setUp().  You might want to figure out how to only restart the mom once.

agrawalravi90 commented:
Thanks guys, I've addressed your comments.

@arungrover run, for PBS_SCP, I'm only reverting it if it is not set. My thinking is that the only test scenario for it is if somebody unsets it to test if PBS uses rcp or not. So, if it is unset, I go through the effort to find the path to it and set PBS_SCP, otherwise i leave it alone. Please let me know if this sounds reasonable.

agrawalravi90 commented:
@arungrover and @bhroam , gentle reminder, thanks

arungrover commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,132 @@ def init_proc_mon(self):
+                    cmds_to_exec.append(["comm", "start"])
+                    dmns_to_start += 1
+            else:
+                pbs_conf_val["PBS_START_MOM"] = "0"
```
why are we turning mom off when we are able to find comm?

agrawalravi90 replied:
Thanks for catching that

arungrover commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,132 @@ def init_proc_mon(self):
+                    cmds_to_exec.append(["mom", "start"])
+                    dmns_to_start += 1
+            else:
+                pbs_conf_val["PBS_START_MOM"] = "0"
```
I am sure we have discussed it, but can you help me jog my memory about why are we disabling mom when self.moms is set. Also, don't you want to append mom stop to cmds_to_exec?

agrawalravi90 replied:
If self.moms is set, we are disabling the mom on the server host as the tests will pick up moms from the self.moms hosts. Does that make sense, or do you think that's a bad idea?

arungrover replied:
But doesn't PTL guarantee a mom on the local host to all the tests? If we do this then many tests will start failing

agrawalravi90 replied:
actually, earlier my code was first turning the local mom off, and then bringing it back up when it parsed the self.moms list. I'm going to just check for it first now and not bring it down unnecessarily

arungrover commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,132 @@ def init_proc_mon(self):
+            self.du.set_pbs_config(mom.hostname, pbs_conf_filepath,
+                                   confs=pbs_conf_val)
+
+            if start_mom:
```
If you change PBS_CORE_LIMITS or PBS_SERVER or PBS_SCP you would still want to restart mom

arungrover commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,132 @@ def init_proc_mon(self):
+                if scppath != "scp":
+                    pbs_conf_val["PBS_SCP"] = scppath
+
+            self.du.set_pbs_config(server.hostname, pbs_conf_filepath,
```
If you change PBS_CORE_LIMITS or PBS_SERVER or PBS_SCP you would still want to restart all the daemons, isn't it?

agrawalravi90 replied:
I think just doing a HUP should suffice as it'll make them do pbs_loadconf() again and read the new values.

arungrover commented on test/tests/selftest/pbs_pbstestsuite.py:
```diff
@@ -0,0 +1,132 @@
+        # Now call revert_pbsconf() & restart PBS again
+        self.server.pi.stop()
+        self.revert_pbsconf()
+        self.server.pi.start()
```
Do you have to do this? isn't revert_pbsconf starting sched daemon automatically?

arungrover commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,132 @@ def init_proc_mon(self):
 
+    def revert_pbsconf(self):
+        """
+        Revert the following about the pbs.conf file:
```
Since the function also restarts daemons when needed, It is better that we mention that in the docstring

agrawalravi90 commented:
@arungrover thanks I've addressed your comments, please provide further feedback

arungrover commented on test/tests/selftest/pbs_pbstestsuite.py:
```diff
@@ -105,28 +109,114 @@ def test_revert_pbsconf_remotemom(self):
+        # First, check that there's no existing core file in mom_priv
+        parent_dir = os.getcwd()
+        snapdirs = []
+        with PBSSnapUtils(out_dir=parent_dir, acct_logs=0, daemon_logs=0,
```
Why do we need to take snapshots? Can we not run an 'ls' command on the mom_priv directory?

agrawalravi90 replied:
we can, but how will we know which file is a core file?

arungrover replied:
I thought they all start with prefix core and if there are already existing core files in the 'priv' directory then daemons store the previous core files by appending an integer value to their name, like core.12434 will become core_001

agrawalravi90 replied:
nope, I've seen core file names start with 'vgcore' or something completely different, it depends on how the system is configured: https://sigquit.wordpress.com/2009/03/13/the-core-pattern/

agrawalravi90 replied:
hmm .. I can see what you are trying to say, why not just go for the simple approach as it might be good enough .. since this is just a PTL test which will most probably run in a controlled environment anyways, I'll change it to the way you are suggesting

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,142 @@ def init_proc_mon(self):
+                mom.signal("-HUP")
+
+        # Revert comm hosts' pbs.conf
+        for comm in self.comms.values():
```
I just had a thought.  You might be setting the same pbs.conf multiple times if self.moms or self.comms share a common host.

bhroam commented on test/tests/selftest/pbs_pbstestsuite.py:
```diff
@@ -0,0 +1,222 @@
+        """
+        pbs_conf_filepath = self.du.get_pbs_conf_file(self.server.hostname)
+        pbs_conf_val = self.du.parse_pbs_config(self.server.hostname,
+                                                pbs_conf_filepath)
```
I just checked, you don't need to call get_pbs_conf_file() before parse_pbs_conf.  If the second argument is not passed to parse_pbs_config(), it will default to calling get_pbs_conf_file(). (and other places in the code)

agrawalravi90 replied:
I'm aware of that, the reason I'm doing it this way is so that it doesn't have to call get_pbs_conf_file() again when I set new pbs.conf attributes inside set_pbs_config()

agrawalravi90 replied:
that's exactly why I was doing it, I guess it won't be too much of an overhead so I'll do what you suggest

bhroam commented on test/tests/selftest/pbs_pbstestsuite.py:
```diff
@@ -0,0 +1,222 @@
+        # Verify that the scheduler came up and start bit is 1
+        self.assertTrue(self.scheduler.isUp())
+        pbs_conf_val = self.du.parse_pbs_config(self.server.hostname)
+        self.assertEqual(pbs_conf_val["PBS_START_SCHED"], "1")
```
Maybe you should check self.server.pbs_conf[]? (and other places in the test)

agrawalravi90 replied:
Since I'm reverting existing pbs.conf values, I didn't want to rely on the cache'd values from self.server.pbs.conf.

agrawalravi90 replied:
oh yes, thanks for catching that, I'll update the cache

bhroam commented on test/tests/selftest/pbs_pbstestsuite.py:
```diff
@@ -0,0 +1,222 @@
+                             (server.hostname))
+
+        # Since the test requested for a mom, the server host's mom should not
+        # be ON, the rest should be set to 1
```
Why does it matter if the server's mom is up or not?  We're checking a remote mom.

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,142 @@ def init_proc_mon(self):
+            start_mom = False
+            pbs_conf_filepath = self.du.get_pbs_conf_file(mom.hostname)
+            pbs_conf_val = self.du.parse_pbs_config(mom.hostname,
+                                                    pbs_conf_filepath)
```
By doing this, you will keep any pbs.conf variables that are set during a test that are not the default values (e.g. PBS_LOG_HIGHRES_TIMESTAMP).  You need to create a dictionary of the variables you want, and set them.

agrawalravi90 replied:
hmm .. ok, this will need some refactoring, so I'll mark this WIP

bhroam commented on test/tests/selftest/pbs_pbstestsuite.py:
```diff
@@ -0,0 +1,222 @@
+    def test_revert_pbsconf_corelimit(self):
+        """
+        Test the functionality of PBSTestSuite.revert_pbsconf() when
+        PBS_CORE_LIMIT is reverted
```
Maybe: 'Test the functionality of PBSTestSuite.revert_pbsconf() when PBS_CORE_LIMIT is set to a value other than the default'

bhroam commented on test/tests/selftest/pbs_pbstestsuite.py:
```diff
@@ -0,0 +1,222 @@
+                for corefile in core_filenames:
+                    pbs_corepath = os.path.join(pbs_conf_val["PBS_HOME"],
+                                                "mom_priv", corefile)
+                    self.du.rm(path=pbs_corepath, force=True, sudo=True)
```
core files are dropped in mom_priv.  This will delete everything in mom's priv directory (like her config file).  Also, I don't think we need to test this.  We aren't testing the mom functionality of dropping a core or not.  We are testing whether or not we revert pbs.conf.

arungrover replied:
Well, it does test that calling revert_pbsconf() makes the changes take effect in daemons. If we do not add this test then we don't know if daemons have been restarted and setup has correctly reverted it.

bhroam replied:
Fair enough.  I can see the logic in that.  How do we know if the daemon has been restarted if we don't test the functionality that has been reverted.  The first part of my comment is still true though.  There are other files in mom_priv other than core files.  We should not delete them.

agrawalravi90 replied:
We aren't deleting anything other than the core files. core_filenames literally is a list of just the core files, and we only delete them from mom_priv.

bhroam replied:
Isn't core_filenames a listdir() of mom_priv?

agrawalravi90 replied:
no, it is a listdir on "core_file_bt/mom_priv", that directory only contains backtraces from core files. I'm going to take Arun's advice and just look for files which start with 'core' instead of taking a snapshot, that should make the test look much more straight-forward

bhroam commented on test/tests/selftest/pbs_pbstestsuite.py:
```diff
@@ -0,0 +1,222 @@
+                                                pbs_conf_filepath)
+        if pbs_conf_val is None or len(pbs_conf_val) < 1:
+            raise ValueError("Could not parse pbs.conf on host %s" %
+                             (server.hostname))
```
This is a test, don't raise exceptions.  Use asserts. (and other places in the test)

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,142 @@ def init_proc_mon(self):
+
+            if dmns_to_start == 4:
+                # Restart all
+                server.restart(server.hostname)
```
Thanks for catching that, I meant to use self.pi.restart(), i even provided it a hostname argument, server.restart() doesn't accept any arguments, I'm surprised that Codacy didn't catch this.

agrawalravi90 commented:
@bhroam and @arungrover I've addressed your comments, made some changes so that the pbs.conf file is reverted to have only the default list of variables, please let me know what you guys think. Thanks.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,294 @@ def init_proc_mon(self):
+        :return default value of the pbs.conf variable if it can be determined
+        as a string, otherwise None
+        """
+        if conf == "PBS_SERVER":
```
the problem is what if the daemons share hosts. For example, if i create a pbs.conf for mom with start bit of only mom set to 1 and the rest set to 0, then it'll be bad if the mom shared that host with the comm or the server daemon.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,294 @@ def init_proc_mon(self):
+                                                    "server", server)
+                    if val is None:
+                        print "Couldn't revert %s in pbs.conf" % (key)
+                        raise
```
when it is raised from the setUp, since we don't catch it, I'm guessing that the test execution will stop and we will bail out of pbs_benchpress. I think this is the right thing to do because if we continue on without a pristine pbs.conf then the results of test execution wouldn't be reliable.

anamikau commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,294 @@ def init_proc_mon(self):
+            # Set PBS_CORE_LIMIT, PBS_SCP and PBS_SERVER
+            if default_vals["PBS_CORE_LIMIT"] != "unlimited":
+                default_vals["PBS_CORE_LIMIT"] = "unlimited"
+                dmns_to_start = 4
```
why are you setting this to 4 explicitly? what if a server do not have a mom on it?

agrawalravi90 replied:
4 is just a way of expressing "restart all daemons on the server". I could have used another flag for this but I didn't think that it was needed. Please let me know if you think that I shouldn't overload dmns_to_start and use another flag instead.

agrawalravi90 replied:
Actually i ended up refactoring it, please let me know if it's still confusing, thanks!

anamikau replied:
Thanks @agrawalravi90. I looked at the updated  code and I am clear now :)

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -840,128 +888,232 @@ def revert_pbsconf(self):
             pbs_conf_val = self.du.parse_pbs_config(server.hostname,
                                                     pbs_conf_filepath)
-            if pbs_conf_val is None or len(pbs_conf_val) < 1:
+            if not pbs_conf_val or len(pbs_conf_val) < 1:
```
I didn't know that an empty list is interpreted as None/False by Python, thanks for the tip

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -840,128 +888,232 @@ def revert_pbsconf(self):
 
+            # to start with, set all keys in default_vals with values from the
+            # existing pbs.conf
+            for conf in default_vals:
```
This is so that we only restart the daemons if necessary. If the existing values are correct then we won't restart the daemons.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -840,128 +888,232 @@ def revert_pbsconf(self):
+            if mom.hostname == primary_server.hostname:
+                continue
+
+            default_vals = none_vals
```
Thanks for catching this!

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -902,30 +898,30 @@ def revert_pbsconf(self):
-            pbs_conf_val = self.du.parse_pbs_config(server.hostname,
-                                                    pbs_conf_filepath)
-            if not pbs_conf_val or len(pbs_conf_val) < 1:
+            self.du.get_pbs_conf_file(server.hostname)
```
Why are you calling this?  It returns the conf file name.  By calling it like this, it does nothing.  parse_pbs_config() will call it itself.

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -989,36 +984,36 @@ def revert_pbsconf(self):
-            pbs_conf_val = self.du.parse_pbs_config(mom.hostname,
-                                                    pbs_conf_filepath)
-            if not pbs_conf_val or len(pbs_conf_val) < 1:
+            self.du.get_pbs_conf_file(mom.hostname)
```
Same as above, this doesn't do anything.

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,282 @@ def init_proc_mon(self):
+                self.du.set_pbs_config(mom.hostname, confs=default_vals,
+                                       append=False)
+                mom.pbs_conf = default_vals
+                mom.pi.initd(mom.hostname, "start", daemon="mom")
```
Don't you want to restart the mom?

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,282 @@ def init_proc_mon(self):
+            if start_comm:
+                self.du.set_pbs_config(comm.hostname, confs=default_vals)
+                comm.pbs_conf = default_vals
+                comm.pi.initd(comm.hostname, "start", daemon="comm")
```
Don't you want to restart the comm?

bhroam commented:
Travis is failing in revert_pbsconf().  You should look into it.

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1012,98 +976,167 @@ def revert_pbsconf(self):
+        Revert contents of the pbs.conf file
+        Also start/stop the appropriate daemons
+        """
+        primary_server = self.servers.values()[0]
```
Is this the primary server, or is self.server the primary server?

agrawalravi90 replied:
If you look at init_servers(), self.server is actually set to servers.values()[0]:
        if cls.servers:
            cls.server = cls.servers.values()[0]

But I think it's probably safer to use self.server as the primary server, so I'll make that change, thanks.

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1012,98 +976,167 @@ def revert_pbsconf(self):
-                self.du.set_pbs_config(comm.hostname, confs=default_vals)
-                comm.pbs_conf = default_vals
-                comm.pi.initd(comm.hostname, "start", daemon="comm")
+        self._revert_pbsconf_server(primary_server, vals_to_set)
```
Why do you call this 3 times?  I don't see any difference between the 3 calls.

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -911,10 +912,13 @@ def _revert_pbsconf_comm(self, primary_server, vals_to_set):
                         new_pbsconf[conf] = val
 
+            for key in keys_to_delete:
+                del(new_pbsconf[key])
```
I don't understand this.  How is this different than deleting the keys as you see them?

agrawalravi90 replied:
Python gets mad at you if you delete elements from a dict while iterating on it:
RuntimeError: dictionary changed size during iteration

bhroam replied:
Now that you say this, it makes perfect sense.  Changing the list/dict you are iterating sounds like a really bad idea.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1012,98 +976,167 @@ def revert_pbsconf(self):
+            if self.moms and server.hostname not in self.moms:
+                if new_pbsconf["PBS_START_MOM"] != "0":
+                    new_pbsconf["PBS_START_MOM"] = "0"
+                    cmds_to_exec.append(["mom", "stop"])
```
Actually, PTL's pi.init.d takes care of this. It creates a temporary file with only the start bit of the relevant daemon set to 1, then calls init script with the operation requested (start/stop) and passes PBS_CONF_FILE set to the temp file. This leaves the original pbs.conf file untouched and still does the right thing with starting/stopping the correct daemon. @saritakh did a great job on this code.

bhroam replied:
Oh cool.  I didn't know that.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -828,6 +829,327 @@ def init_proc_mon(self):
+                                                    primary_server.hostname,
+                                                    "comm", comm)
+                    if val is None:
+                        self.logger.error("Couldn't set %s in pbs.conf" %
```
> Using couldn't feels like that the program tried but it couldn't make it

Well, that actually is what's happening though, we try to find the default value for this variable in _get_dflt_pbsconfval(), but if that returns None, we say that we tried but couldn't figure out how to do it. Does it make sense now?

arungrover replied:
Well, I'm ok if you want to retain it the way it is now. I thought, in this case, the log could be that it couldn't figure out a default value for the configuration. 

To me, the current log message just feels like a thing to say when setup tried putting the configuration in the conf file but couldn't do so because of factors like unable to open the file or file not present or write failed etc.

agrawalravi90 replied:
ok, i see what you are saying, alright I'll change it to say "Couldn't revert %s in pbs.conf to its default value"

agrawalravi90 commented:
@arungrover and @bhroam Thanks for all the comments, I've addressed them. I also, finally, ran the remote mom PTL test and fixed some minor issues to make it work. Please let me know if things look okay. Latest test logs attached in the description.

agrawalravi90 commented:
Thanks @bhroam 
@arungrover please let me know if things look okay to you as well. Thanks!

arungrover commented:
looks good @agrawalravi90 please squash and rebase

agrawalravi90 commented:
Thanks a lot @arungrover , I've rebased, hopefully the CI will pass, please merge this in if possible. Thanks again.
